### PR TITLE
aix,ibmi: use uv_interface_addresses instead of getifaddrs

### PR DESCRIPTION
**Pull request by abmusse**
Use uv_interface_addresses to get the interfaces addresses on AIX and IBM i as ifaddrs is not available on these platforms.
After retrieving the interfaces we can then set the ipv6 link local scope id.

https://github.com/libuv/libuv/blob/54d8364c2406758b572621af381f1d83e01ae46c/src/unix/tcp.c#L226-L248

AIX does not have `getifaddrs` but we do have code in `uv_interface_addresses` to get the interface addresses.

https://github.com/libuv/libuv/blob/f01219dfb716ceac9cc7dbc70022a197b20d27b3/src/unix/aix.c#L1116-L1293

Looks like `uv_interface_addresses` uses getifaddrs on Linux

https://github.com/libuv/libuv/blob/f01219dfb716ceac9cc7dbc70022a197b20d27b3/src/unix/linux.c#L1927-L2012

@bnoordhuis is there a reason why we shouldn't use uv_interface_addresses in `uv__ipv6_link_local_scope_id`?

CC
@richardlau 

Part of https://github.com/libuv/libuv/issues/4117

### CODE REVIEW

**Comment by bnoordhuis** [comment]
> is there a reason why we shouldn't use uv_interface_addresses in uv__ipv6_link_local_scope_id?

Yes, unnecessary extra memory allocations. Both getifaddrs() and uv_interface_addresses() malloc.

**Comment by richardlau** [comment]
> Ran Test Build on AIX CI:
> 
> https://ci.nodejs.org/job/libuv-test-commit-aix/nodes=aix72-ppc64/2275/console
> 
> ```
> not ok 309 - tcp_connect6_link_local
> # exit code 393222
> # Output from process `tcp_connect6_link_local`:
> # Assertion failed in test/test-tcp-connect6-error.c on line 102: `uv_tcp_connect(&req, &server, (struct sockaddr*) &addr, connect_cb) == 0` (-68 == 0)
> ```
> 
> 68 being returned here looks to map to EADDRNOTAVAIL

FWIW from the same job the only interface on the machine with an IPv6 address is lo0:
```console
# uv_interface_addresses:
#   name: en0
#   internal: 0
#   physical address: 2a:b0:26:58:5b:02
#   address: 10.1.0.6
#   netmask: 255.255.254.0
#   name: en1
#   internal: 0
#   physical address: 2a:b0:26:58:5b:03
#   address: 140.211.9.101
#   netmask: 255.255.255.0
#   name: lo0
#   internal: 1
#   physical address: 00:00:00:00:00:00
#   address: 127.0.0.1
#   netmask: 255.0.0.0
#   name: lo0
#   internal: 1
#   physical address: 00:00:00:00:00:00
#   address: ::1
#   netmask: ffff:ffff:ffff:ffff::
```

File: src/unix/tcp.c [hunk 1]
```diff
@@ -243,6 +248,29 @@ static int uv__ipv6_link_local_scope_id(void) {
...
+  int count, i;
+
+  if (uv_interface_addresses(&interfaces, &count)) {
+    return 0;
+  }
+
+  rv = 0;
+
+  for (i = 0; i < count; i++) {
+    if (uv__is_ipv6_link_local((const struct sockaddr*) &interfaces[i].address.address6)) {
```
Comment (richardlau) [inline src/unix/tcp.c:262]: You'll need to guard this on whether `interfaces[i].address->sa_family == AF_INET6` to avoid misinterpreting the address.

Comment (richardlau) [reply src/unix/tcp.c:262]: hmm although now I see `uv__is_ipv6_link_local` already checks.

File: src/unix/tcp.c [hunk 2]
```diff
@@ -30,13 +30,15 @@
 #include <sys/types.h>
 #include <sys/socket.h>
 
+#if !defined(_AIX)
```
Comment (bnoordhuis) [inline src/unix/tcp.c:33]:
```suggestion
#if !defined(_AIX) && !defined(__PASE__)
```
It doesn't have ifaddrs.h either, or does it?

File: src/unix/tcp.c [hunk 3]
```diff
@@ -225,9 +227,33 @@ static int uv__is_ipv6_link_local(const struct sockaddr* addr) {
 
 static int uv__ipv6_link_local_scope_id(void) {
   struct sockaddr_in6* a6;
+  int rv;
+  // AIX path fallback to use uv_interface_addresses
+#if defined (_AIX) && !defined(__PASE__)
```
Comment (bnoordhuis) [inline src/unix/tcp.c:232]: No C99 comments and if you're going to effectively replace the function with another then for legibility reasons just do so. That is:
```c
static int uv__ipv6_link_local_scope_id(void) {
#if defined (_AIX) && !defined(__PASE__)
  // ...
#else
  // ...
#endif  /* defined (_AIX) && !defined(__PASE__) */
}
```

File: src/unix/tcp.c [hunk 4]
```diff
@@ -225,9 +227,33 @@ static int uv__is_ipv6_link_local(const struct sockaddr* addr) {
...
+  uv_interface_address_t* interfaces;
+  int count, i;
+
+  if (uv_interface_addresses(&interfaces, &count)) {
+    return 0;
+  }
```
Comment (bnoordhuis) [inline src/unix/tcp.c:238]: Style:
```suggestion
  if (uv_interface_addresses(&interfaces, &count))
    return 0;
```

File: src/unix/tcp.c [hunk 5]
```diff
@@ -225,9 +227,33 @@ static int uv__is_ipv6_link_local(const struct sockaddr* addr) {
...
+
+  for (i = 0; i < count; i++) {
+    if (uv__is_ipv6_link_local((const struct sockaddr*) &interfaces[i].address.address6)) {
+      if (&interfaces[i].address.address6 != NULL) {
+        a6 = &interfaces[i].address.address6;
+        rv = a6->sin6_scope_id;
+      }
+      break;
+    }
+  }
```
Comment (bnoordhuis) [inline src/unix/tcp.c:250]:
```suggestion
  for (ifa = interfaces; ifa != &interfaces[count]; ifa++) {
    if (uv__is_ipv6_link_local((struct sockaddr*) &ifa->address)) {
      rv = ifa->address.address6.sin6_scope_id;
      break;
    }
  }
```

Comment (bnoordhuis) [reply src/unix/tcp.c]: The errors suggest you have `uv_interface_address_t` (value type), not `uv_interface_address_t*` (pointer type)

File: src/unix/tcp.c [hunk 6]
```diff
@@ -225,9 +225,31 @@ static int uv__ipv6_link_local_scope_id(void) {
   return 0;
 #else
   struct sockaddr_in6* a6;
+  int rv;
+#if defined(_AIX)
```
Comment (bnoordhuis) [inline src/unix/tcp.c:229]: This is dead code because of the `#if defined(_AIX)` on line 224.

**Review by bnoordhuis — APPROVED** [review]
LGTM, thanks!

**Comment by bnoordhuis** [comment]
One question: the comments and the commit log say "aix _and_ ibmi" but the ifdef guard only checks AIX. Is this PR just for AIX or for both platforms?

**Pull request merged**